crates/autopilot/src/run_loop.rs

```diff
@@ -450,45 +408,77 @@ impl RunLoop {
-                    Metrics::solve_err(driver, start.elapsed(), &err);
-                    if matches!(err, SolveError::NoSolutions) {
-                        tracing::debug!(driver = %driver.name, "solver found no solution");
-                    } else {
-                        tracing::warn!(?err, driver = %driver.name, "solve error");
-                    }
-                    vec![]
-                }
-            };
-
-            solutions.into_iter().filter_map(|solution| match solution {
-                Ok(solution) => {
-                    Metrics::solution_ok(driver);
-                    Some(Participant { driver, solution })
-                }
-                Err(err) => {
-                    Metrics::solution_err(driver, &err);
-                    tracing::debug!(?err, driver = %driver.name, "invalid proposed solution");
-                    None
-                }
-            })
-        }))
+        let mut solutions = futures::future::join_all(
+            self.drivers
+                .iter()
+                .map(|driver| self.solve(driver, request)),
+        )
         .await
         .into_iter()
         .flatten()
-        .collect()
+        .collect::<Vec<_>>();
+
+        // Shuffle so that sorting randomly splits ties.
+        solutions.shuffle(&mut rand::thread_rng());
+        solutions.sort_unstable_by_key(|participant| participant.solution.score().get().0);
+
+        // Make sure the winning solution is fair.
+        while !Self::is_solution_fair(solutions.last(), &solutions, auction) {
+            let unfair_solution = solutions.pop().expect("must exist");
+            warn!(
+                invalidated = unfair_solution.driver.name,
+                "fairness check invalidated of solution"
+            );
+        }
+        self.report_on_solutions(&solutions, auction);
+
+        solutions
+    }
+
+    /// Records metrics, order events and logs for the given solutions.
+    /// Expects the winning solution to be the last in the list.
+    fn report_on_solutions(&self, solutions: &[Participant<'_>], auction: &domain::Auction) {
+        let Some(winner) = solutions.last() else {
+            // no solutions means nothing to report
+            return;
+        };
+
+        solutions.iter().for_each(|solution| {
+            tracing::debug!(
+                driver=%solution.driver.name,
+                orders=?solution.solution.order_ids(),
+                "proposed solution"
+            );
+        });
+
+        let proposed_orders: HashSet<_> = solutions
+            .iter()
+            .flat_map(|solution| solution.solution.order_ids().copied())
+            .collect();
+        let winning_orders: HashSet<_> = solutions
+            .last()
+            .into_iter()
+            .flat_map(|solution| solution.solution.order_ids().copied())
+            .collect();
+        let mut non_winning_orders: HashSet<_> = proposed_orders
+            .difference(&winning_orders)
```

Comment:
it would be faster if we just get the vector from 0 to `last - 2`, instead of calculating the difference :thinking:

Reply:
The idea is to only count orders that have not also been proposed by the winning solution. In case a non-winning solution would also solve an order that the winning solution solved your approach would produce false positives.

Reply:
Oh, I understand, thanks!